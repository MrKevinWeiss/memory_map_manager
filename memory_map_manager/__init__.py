"""memory_map_manager package."""
__version__ = "0.1.0"
