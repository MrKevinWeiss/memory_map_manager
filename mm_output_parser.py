#!/usr/bin/env python3
"""This module generates output files based on memory maps."""
from gen_helpers import to_camel_case, to_underscore_case


def _find_unique_keys(mem_map):
    field_names = []
    for record in mem_map:
        for field_name in record.keys():
            if field_name not in field_names:
                field_names.append(field_name)
    field_names = sorted(field_names)
    field_names.insert(0, field_names.pop(field_names.index('description')))
    field_names.insert(0, field_names.pop(field_names.index('total_size')))
    field_names.insert(0, field_names.pop(field_names.index('offset')))
    field_names.insert(0, field_names.pop(field_names.index('name')))
    return field_names


def parse_mem_map_to_csv(mem_map, name_delimiter='.'):
    """Parses a memory map to a csv table string."""
    fields = _find_unique_keys(mem_map)
    csv_str = ','.join(fields)
    for record in mem_map:
        csv_str += '\n'
        name = name_delimiter.join(record['name'])
        for field in fields:
            if field not in record:
                record[field] = ''
            if record[field] is None:
                record[field] = ''
            if field == 'name':
                csv_str += name
            else:
                csv_str += str(record[field]).replace(',', '","')
            csv_str += ','
        csv_str = csv_str[:-1]
    return csv_str


def _record_to_getter(record, fxn_call_list=None):
    fxn_str = 'get_{}'.format('_'.join(record['name']))
    getter_str = "    def {}(self):\n".format(fxn_str)
    getter_str += "        \"\"\"{}\"\"\"\n".format(record['description'])
    if record['is_bitfield']:
        getter_str += "        return self.read_bits"
        getter_str += "({}, {}, {})\n".format(record['offset'],
                                              record['bit_offset'],
                                              record['bits'])
    else:
        getter_str += "        return self.read_bytes"
        getter_str += "({}, {})\n".format(record['offset'],
                                          record['total_size'])
    if isinstance(fxn_call_list, list):
        fxn_call_list.append(fxn_str)
    return getter_str


def _record_to_setter(record, fxn_call_list=None):
    fxn_str = 'set_{}'.format('_'.join(record['name']))
    setter_str = "    def {}(self, data={}):\n".format(fxn_str,
                                                       record['default'])
    setter_str += "        \"\"\"{}\"\"\"\n".format(record['description'])
    if record['is_bitfield']:
        setter_str += "        return self.write_bits"
        setter_str += "({}, {}, {}, data)\n".format(record['offset'],
                                                    record['bit_offset'],
                                                    record['bits'])
    else:
        setter_str += "        return self.write_bytes"
        setter_str += "({}, data, {})\n".format(record['offset'],
                                                record['total_size'])
    if isinstance(fxn_call_list, list):
        fxn_call_list.append(fxn_str)
    return setter_str


def parse_mem_map_to_if(mem_map, class_name, parent_module=None):
    """Parses a memory map to a python interface to a .py string."""
    if_str = ""
    if parent_module is not None:
        if_str += "from {} ".format(to_underscore_case(parent_module))
        if_str += "import {}\n\n\n".format(to_camel_case(parent_module))
    if_str += "class {}".format(to_camel_case(class_name))
    if parent_module is not None:
        if_str += "({})".format(to_camel_case(parent_module))
    if_str += ':\n'
    fxn_call_list = []
    for record in mem_map:
        if_str += _record_to_getter(record, fxn_call_list)
        if_str += '\n'
        if record['access'] != 0:
            if_str += _record_to_setter(record, fxn_call_list)
            if_str += '\n'

    if_str += "    def get_command_list(self):\n"
    if_str += "        \"\"\"A list of all possible commands\"\"\"\n"
    if_str += "        cmds = list()\n"
    for fxn_call in fxn_call_list:
        if_str += "        cmds.append(self.{})\n".format(fxn_call)
    if_str += "        return cmds\n"
    return if_str


def parse_mem_map_to_access_c(mem_map):
    """Parses access registers based on memory map to a .c string."""
    a_str = "#include \"app_access.h\"\n\n"
    a_str += "const uint8_t REG_ACCESS[] = { \n"
    size = 0
    for record in mem_map:
        if record['is_bitfield'] is False:
            for access_byte in range(record["total_size"]):
                if access_byte != 0:
                    a_str += ", "
                a_str += "0x%02X" % record["access"]
                size += 1
            if record != mem_map[-1]:
                a_str += ","
            a_str += " /* {} */\n".format('_'.join(record["name"]))
    a_str = a_str.rstrip(',')
    a_str += "/* total size %d */\n};" % size
    return a_str


def main():
    """Tests parsing each output based on example typedef."""
    from td_parser import parse_basic_typedefs
    from mm_parser import parse_typedefs_to_mem_map
    from gen_helpers import TEST_T

    typedefs = parse_basic_typedefs(TEST_T)
    mem_map = parse_typedefs_to_mem_map(typedefs)
    mem_map[0]['default'] = 6

    base_if = parse_mem_map_to_if(mem_map, 'base_if')
    test_if = parse_mem_map_to_if(mem_map, 'test_if', 'base_if')
    access_c = parse_mem_map_to_access_c(mem_map)

    print(base_if)
    print(test_if)
    print(access_c)


if __name__ == "__main__":
    main()
