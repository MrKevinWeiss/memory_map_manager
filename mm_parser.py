#!/usr/bin/env python3
"""Parses and manages memory maps from typedefs"""
import copy
from gen_helpers import PRIM_TYPES


def _copy_elements(typename, typedefs):
    for typedef in typedefs:
        if typename == typedef["name"]:
            return {'elements': copy.deepcopy(typedef["elements"])}
    raise ValueError("Cannot find {}".format(typename))


def _copy_typedef_elements(element, expanded_typedefs):
    e_type = element["type"]
    if "array_size" in element:
        a_size = element["array_size"]
        element['array'] = [_copy_elements(e_type, expanded_typedefs)
                            for i in range(a_size)]
    else:
        element.update(_copy_elements(e_type, expanded_typedefs))


def _expand_typedefs(typedefs):
    expanded_typedefs = []
    for typedef in typedefs:
        for element in typedef["elements"]:
            if element["type"] not in PRIM_TYPES:
                if "bitfield" not in element:
                    _copy_typedef_elements(element, expanded_typedefs)
        expanded_typedefs.append(typedef)
    return expanded_typedefs


def _update_offsets(elements, offset=0):
    for element in elements:
        element["offset"] = offset
        if "elements" in element:
            offset = _update_offsets(element["elements"], offset)
        elif "array" in element:
            for array_val in element["array"]:
                offset = _update_offsets(array_val["elements"], offset)
        elif "array_size" in element:
            offset += element["size"]*element["array_size"]
        else:
            offset += element["size"]
    return offset


def _expand_mem_map(typedefs, mem_map=None):

    exp_typedefs = _expand_typedefs(typedefs)
    if mem_map is None:
        mem_map = exp_typedefs[-1]
    else:
        mem_map = next(itm for itm in exp_typedefs if itm["name"] == "mem_map")
    _update_offsets(mem_map['elements'])
    return mem_map


def _element_to_bitfield_record(name, element):

    bitfields = []
    offset = 0
    for bit_info in element["bitfield"]:
        bitfield = {}
        bitfield = copy.deepcopy(element)
        bitfield["type"] = bitfield.pop("bit_type")
        bitfield["total_size"] = element["size"]
        bitfield["is_bitfield"] = True
        bitfield["bits"] = bit_info["bits"]
        bitfield["description"] = bit_info["description"]
        bitfield["bit_offset"] = offset
        bitfield.pop('bitfield', None)
        name.append(bit_info["name"])
        bitfield["name"] = copy.deepcopy(name)
        name.pop()
        offset += bit_info["bits"]
        bitfields.append(bitfield)
    return bitfields


def _element_to_mem_map_record(name, element, mem_map):
    mem_map.append(copy.deepcopy(element))
    name.append(element["name"])
    if "array_size" in element:
        mem_map[-1]["total_size"] = element["size"] * element["array_size"]
        name.append("%d" % element["array_size"])
    else:
        mem_map[-1]["total_size"] = element["size"]
    mem_map[-1].pop('array_size', None)
    mem_map[-1].pop('bitfield', None)
    mem_map[-1].pop('bit_type', None)

    mem_map[-1]["name"] = copy.deepcopy(name)
    mem_map[-1]["bits"] = element["size"] * 8
    mem_map[-1]["bit_offset"] = 0
    mem_map[-1]["is_bitfield"] = False

    if 'bitfield' in element:
        mem_map.extend(_element_to_bitfield_record(name, element))
    if "array_size" in element:
        name.pop()
    name.pop()


def _parse_elements_to_mem_map(elements, mem_map=None, name=None):
    if mem_map is None:
        mem_map = []
    if name is None:
        name = []
    for element in elements:
        if "elements" in element:
            name.append(element["name"])
            _parse_elements_to_mem_map(element["elements"], mem_map, name)
            name.pop()
        elif "array" in element:
            name.append(element["name"])
            for i, array_val in enumerate(element["array"]):
                name.append("%d" % i)
                _parse_elements_to_mem_map(array_val["elements"],
                                           mem_map, name)
                name.pop()
            name.pop()
        else:
            _element_to_mem_map_record(name, element, mem_map)
    return mem_map


def parse_typedefs_to_mem_map(typedefs, mem_map=None):
    """Parses a selected (or the last) typedef to a memory map"""
    mem_map_expanded = _expand_mem_map(typedefs, mem_map=None)
    mem_map = _parse_elements_to_mem_map(mem_map_expanded['elements'])
    return mem_map


def import_mem_map_values(mem_map, saved_mem_map,
                          generated_fields=None):
    """Imports type_name values from saved memory maps"""
    if generated_fields is None:
        generated_fields = ['bit_offset', 'bits', 'is_bitfield', 'bit_offset',
                            'name', 'offset', 'size', 'total_size', 'type']
    for saved_record in saved_mem_map:
        for record in mem_map:
            if record['name'] == saved_record['name']:
                for field_name in saved_record.keys():
                    if field_name not in generated_fields:
                        record[field_name] = saved_record[field_name]


def main():
    """Tests the parsing and updating saved values from the memory map."""
    from pprint import pprint
    from gen_helpers import TEST_T
    import td_parser

    typedefs = td_parser.parse_basic_typedefs(TEST_T)

    mem_map = parse_typedefs_to_mem_map(typedefs)
    saved_mem_map = copy.deepcopy(mem_map)
    saved_mem_map[0]['access'] = 99
    saved_mem_map[1]['default'] = '123'
    saved_mem_map[1]['description'] = 'test to alter description'

    print("================ MEM MAP =====================\n")
    pprint(mem_map, width=120)
    print("=============== SAVED MEM MAP ================\n")
    pprint(saved_mem_map, width=120)

    print('============ Diff ============')
    try:
        from deepdiff import DeepDiff
        pprint(DeepDiff(mem_map, saved_mem_map))
    except ImportError:
        print("Cannot make diff, try 'pip install deepdiff'")

    import_mem_map_values(mem_map, saved_mem_map)
    print("============== MODIFIED MEM MAP ===============\n")
    pprint(mem_map, width=120)

    print('============ Diff After Modified ============')
    try:
        from deepdiff import DeepDiff
        pprint(DeepDiff(mem_map, saved_mem_map))
    except ImportError:
        print("Cannot make diff, try 'pip install deepdiff'")


if __name__ == "__main__":
    main()
