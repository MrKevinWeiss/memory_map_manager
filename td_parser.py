#!/usr/bin/env python3
"""Fills in any missing/calculatable information of typedefs"""
from copy import deepcopy
from gen_helpers import PRIM_TYPES


def _fill_res_element(typedef, total_byte):
    res = {'name': 'res',
           'type': 'uint8_t',
           'array_size': typedef['size'] - total_byte,
           'size': 1,
           'description': 'Reserved bytes',
           'access': 0x00,
           'default': None}
    typedef["elements"].append(res)


def _update_typedef_size(typedef, total_byte):
    if 'size' in typedef:
        if typedef['size'] < total_byte:
            raise ValueError("{} to large".format(typedef["name"]))

        if typedef['size'] is not total_byte:
            _fill_res_element(typedef, total_byte)
    else:
        typedef['size'] = total_byte


def _update_element_sizes(elements, type_sizes):
    total_byte = 0
    for element in elements:
        if 'bitfield' in element:
            element["size"] = PRIM_TYPES[element["bit_type"]]
        else:
            element["size"] = type_sizes[element["type"]]

        if 'array_size' in element:
            total_byte += element["size"] * element["array_size"]
        else:
            total_byte += element["size"]
    return total_byte


def _update_typedef_sizes(typedefs):
    type_sizes = {}
    type_sizes.update(PRIM_TYPES)
    for typedef in typedefs:
        total_byte = _update_element_sizes(typedef["elements"], type_sizes)
        _update_typedef_size(typedef, total_byte)
        type_sizes[typedef["name"]] = typedef["size"]


def _fill_elements(elements, type_name, value, bitfield):
    for element in elements:
        if type_name not in element:
            element[type_name] = value
        if bitfield and 'bitfield' in element:
            for bf_val in element['bitfield']:
                if type_name not in bf_val:
                    bf_val[type_name] = value


def _fill_empty_typedefs(typedefs, type_name, value,
                         overwrite=True, bitfield=False):
    for typedef in typedefs:
        if type_name not in typedef:
            typedef[type_name] = value

        if overwrite:
            _fill_elements(typedef['elements'], type_name, value, bitfield)
        else:
            _fill_elements(typedef['elements'], type_name,
                           typedef[type_name], bitfield)


def parse_basic_typedefs(typedefs):
    """Fills in any missing/calculatable information of typedefs."""
    # Calculates sizes and res bytes
    cp_typedefs = deepcopy(typedefs)
    _update_typedef_sizes(cp_typedefs)
    _fill_empty_typedefs(cp_typedefs, 'description', '', bitfield=True)
    _fill_empty_typedefs(cp_typedefs, 'default', None)
    _fill_empty_typedefs(cp_typedefs, 'access', 1, False)
    return cp_typedefs


def main():
    """Prints example of the parsing."""
    from pprint import pprint
    from gen_helpers import TEST_T

    print('============ Before Parsing ============')
    pprint(TEST_T)
    typedefs = parse_basic_typedefs(TEST_T)
    print('============ After Parsing ============')
    pprint(typedefs)
    print('============ Diff ============')
    try:
        from deepdiff import DeepDiff

        pprint(DeepDiff(TEST_T, typedefs))
    except ImportError:
        print("Cannot make diff, try 'pip install deepdiff'")


if __name__ == "__main__":
    main()
