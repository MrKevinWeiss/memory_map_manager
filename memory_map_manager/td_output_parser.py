#!/usr/bin/env python3
"""This module generates output files based on typedefs."""


def _metadata_to_h_intro(metadata):
    intro_str = "/*\n"
    intro_str += " * Filename: {}.g\n".format(metadata["name"])
    intro_str += " * Author: {}\n".format(metadata["author"])
    intro_str += " * Revision: {}\n".format(metadata["revision"])
    intro_str += " */\n\n"
    intro_str += "#ifndef %s_H_\n" % (metadata["name"].upper())
    intro_str += "#define %s_H_\n" % (metadata["name"].upper())
    return intro_str


def _bitfield_to_c_struct(bf_element):
    bf_str = ""
    bf_str += "/* @brief {} */\n".format(bf_element["description"])
    bf_str += "typedef struct %s_TAG {\n" % (bf_element["type"])
    total_bits = 0
    for val in bf_element["bitfield"]:
        bf_str += "\t/* {} */\n".format(val["description"])
        bf_str += "\t{} {} : {};\n".format(bf_element["bit_type"],
                                           val["name"], val["bits"])
        total_bits += int(val["bits"])
    bf_str += "} %s;\n\n" % (bf_element["type"])

    if bf_element["size"] < ((float(total_bits)/8)):
        raise ValueError("Too many bits in %s for %s" %
                         (bf_element["type"], bf_element["bit_type"]))
    return bf_str


def _typedef_to_c_struct(typedef):
    c_str = ""
    for element in typedef["elements"]:
        if 'bitfield' in element:
            c_str += _bitfield_to_c_struct(element)
    c_str += "/* @brief {} */\n".format(typedef["description"])
    c_str += "typedef union %s_TAG {\n" % (typedef["name"])
    c_str += "\tstruct {\n"

    for element in typedef["elements"]:
        c_str += "\t\t/* {} */\n".format(element["description"])
        if 'array_size' in element:
            c_str += "\t\t{} {}[{}];\n".format(element["type"],
                                               element["name"],
                                               element["array_size"])
        else:
            c_str += "\t\t{} {};\n".format(element["type"], element["name"])
    c_str += "\t};\n"
    c_str += "\tuint8_t data8[{}];\n".format(typedef["size"])
    c_str += "} %s;\n" % (typedef["name"])
    return c_str


def parse_typedefs_to_h(typedefs, metadata=None):
    """Parses the typedef to a c header containing typedef structs."""
    td_str = ""
    if metadata is not None:
        td_str += _metadata_to_h_intro(metadata)
    td_str += "\n"
    td_str += "#include <stdint.h>\n"
    td_str += "\n"
    td_str += "#pragma pack(1)\n"
    for typedef in typedefs:
        td_str += _typedef_to_c_struct(typedef)
        td_str += "\n"
    td_str += "#pragma pack()\n"
    if metadata is not None:
        td_str += "#endif"
    return td_str


def main():
    """Tests parsing example typedef to c header."""
    from td_parser import parse_basic_typedefs
    from gen_helpers import TEST_T
    from gen_helpers import TEST_MD

    typedefs = parse_basic_typedefs(TEST_T)
    print(parse_typedefs_to_h(typedefs, TEST_MD))


if __name__ == "__main__":
    main()
