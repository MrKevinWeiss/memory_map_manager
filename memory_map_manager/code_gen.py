#!/usr/bin/env python3
"""This module handles parsing of typedefs and creating outputs"""
import os
import json
import argparse
from td_parser import parse_basic_typedefs
from mm_parser import parse_typedefs_to_mem_map, import_mem_map_values
from td_output_parser import parse_typedefs_to_h
from mm_output_parser import parse_mem_map_to_access_c, parse_mem_map_to_csv
from mm_output_parser import parse_mem_map_to_if
from gen_helpers import to_underscore_case


def _parse_filename(f_arg, d_arg, name_contains):
    if f_arg is not None:
        for file in os.listdir(os.path.join(os.path.dirname(__file__), d_arg)):
            if file.endswith(".json"):
                if name_contains in file:
                    return file
    return None


def main():
    """Parses typedefs and creates outputs"""

    parser = argparse.ArgumentParser()
    parser.add_argument("--input_dir",
                        help='Directory where all input files are stored',
                        default='saved_setups/')
    parser.add_argument("--output_dir",
                        help='Output directory where results are stored',
                        default='output/')
    parser.add_argument("--import_td_f",
                        help='Typedef file to import and parse')
    parser.add_argument("--import_mm_f",
                        help='Saved memory map file to not '
                        'overwrite changes to access, '
                        'description, and default values')
    parser.add_argument("--output_mm_f",
                        help='Output file for the memory map')
    parser.add_argument("--csv_name",
                        help='Output filename of the memory map csv',
                        default='mem_map.csv')
    parser.add_argument("--td_h_name",
                        help='Output filename of the typedef c header',
                        default='app_typedef.h')
    parser.add_argument("--access_name",
                        help='Output filename of the access.c file',
                        default='app_access.c')
    parser.add_argument("--py_if_name",
                        help='Output filename of python interface')
    parser.add_argument("--if_parent",
                        help='Name of the parent class the python interface '
                             'is based on',
                        default="BaseDevice")
    args = parser.parse_args()

    import_td_f = _parse_filename(args.import_td_f, args.input_dir, "typedef")
    import_mm_f = _parse_filename(args.import_mm_f, args.input_dir, "mem_map")

    with open(os.path.join(args.input_dir, import_td_f)) as td_f:
        json_data = json.load(td_f)
    typedefs = json_data['typedefs']
    metadata = json_data['metadata']

    typedefs = parse_basic_typedefs(typedefs)
    mem_map = parse_typedefs_to_mem_map(typedefs)

    if import_mm_f is not None:
        with open(os.path.join(args.input_dir, import_mm_f)) as mm_f:
            imported_mem_map = json.load(mm_f)
        import_mem_map_values(mem_map, imported_mem_map)
    else:
        import_mm_f = 'mem_map.json'

    if not os.path.exists(args.input_dir):
        os.makedirs(args.input_dir)
    if import_mm_f is None:
        import_mm_f = import_td_f.replace("typedef", "mem_map")
    with open(os.path.join(args.input_dir, import_mm_f), 'w') as outfile:
        json.dump(mem_map, outfile, indent=4, sort_keys=True)

    if not os.path.exists(args.output_dir):
        os.makedirs(args.output_dir)

    with open(os.path.join(args.output_dir, args.csv_name), 'w') as outfile:
        outfile.write(parse_mem_map_to_csv(mem_map))

    with open(os.path.join(args.output_dir, args.access_name), 'w') as outfile:
        outfile.write(parse_mem_map_to_access_c(mem_map))

    if args.py_if_name is None:
        args.py_if_name = to_underscore_case(metadata['name'])
    with open(os.path.join(args.output_dir, args.py_if_name + '.py'),
              'w') as outfile:
        outfile.write(parse_mem_map_to_if(mem_map, args.py_if_name,
                                          args.if_parent))

    with open(os.path.join(args.output_dir, args.td_h_name), 'w') as outfile:
        outfile.write(parse_typedefs_to_h(typedefs, metadata))


if __name__ == "__main__":
    main()
