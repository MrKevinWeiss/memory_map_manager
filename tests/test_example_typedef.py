# Copyright (c) 2018 Kevin Weiss, for HAW Hamburg  <kevin.weiss@haw-hamburg.de>
#
# This file is subject to the terms and conditions of the MIT License. See the
# file LICENSE in the top level directory for more details.
# SPDX-License-Identifier:    MIT
"""Tests Serial Driver implmentation in RIOT PAL."""
from copy import deepcopy
import json
import difflib
import pytest
from memory_map_manager.config_import_export import import_config
from memory_map_manager.td_parser import update_typedefs
from memory_map_manager.mm_parser import parse_typedefs_to_mem_maps
from memory_map_manager.td_output_parser import parse_typedefs_to_h
from memory_map_manager.mm_output_parser import parse_mem_map_to_defaults_h
from memory_map_manager.mm_output_parser import parse_mem_map_to_map_v_c
from memory_map_manager.mm_output_parser import parse_mem_map_to_map_v_h
from memory_map_manager.mm_output_parser import parse_mem_map_to_defaults_c
from memory_map_manager.mm_output_parser import parse_mem_map_to_csv
from memory_map_manager.mm_output_parser import parse_mem_map_to_access_c


@pytest.fixture
def example_config():
    yield import_config('example_typedef.json')


# pylint: disable=W0621
def test_import_regression(regtest, example_config):
    """Tests the import from json against accepted input"""
    regtest.write(json.dumps(example_config, sort_keys=True, indent=4))


# pylint: disable=W0621
def test_updated_regression(regtest, example_config):
    """Tests the full transform of the config file against accepted value"""
    update_typedefs(example_config)
    parse_typedefs_to_mem_maps(example_config, True)
    regtest.write(json.dumps(example_config, sort_keys=True, indent=4))


# pylint: disable=W0621
def test_diff_generated_imported(regtest, example_config):
    """Tests to see if importing non-generated values works"""
    update_typedefs(example_config)
    generated_config = deepcopy(example_config)
    parse_typedefs_to_mem_maps(example_config, True)
    parse_typedefs_to_mem_maps(generated_config, False)
    imp_mm = json.dumps(example_config['mem_maps'], sort_keys=True, indent=4)
    gen_mm = json.dumps(generated_config['mem_maps'], sort_keys=True, indent=4)
    for outstr in difflib.unified_diff(imp_mm.splitlines(),
                                       gen_mm.splitlines(), n=0):
        last_str = outstr
    regtest.write(last_str)


# pylint: disable=W0621
def test_c_h_outputs(regtest, example_config):
    """Test to generated c and h stay the same."""
    update_typedefs(example_config)
    parse_typedefs_to_mem_maps(example_config, False)
    regtest.write(parse_typedefs_to_h(example_config))
    # regtest.write seems to have a bug with the rendering of 0xn, 0xm
    # as a result it is expected that we get something like 0x00, 0x????????
    regtest.write(parse_mem_map_to_access_c(example_config))
    regtest.write(parse_mem_map_to_map_v_c(example_config))
    regtest.write(parse_mem_map_to_map_v_h(example_config))
    regtest.write(parse_mem_map_to_defaults_h(example_config))
    regtest.write(parse_mem_map_to_defaults_c(example_config))


# pylint: disable=W0621
def test_csv_outputs(regtest, example_config):
    update_typedefs(example_config)
    parse_typedefs_to_mem_maps(example_config, True)
    for mem_map in example_config['mem_maps']:
        regtest.write(parse_mem_map_to_csv(mem_map))
