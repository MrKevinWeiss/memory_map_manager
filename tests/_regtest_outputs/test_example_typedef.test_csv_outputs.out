name,offset,total_size,type_size,type,description,access,array_size,bit_offset,bits,default,invalid,max,min
an_element.basic_element,0,,1,uint8_t,,1,,,,,,,
an_element.access_3_element,1,,4,uint32_t,,3,,,,,,,
an_element.access_0_element,5,,4,uint32_t,,0,,,,,,,
an_element.specified_element,9,,4,uint32_t,,1,,,,1,11,10,0
an_element.array_element,13,36,4,int32_t,,1,9,,,,,,
an_element.other_element,49,,2,uint16_t,,1,,,,,,,
an_element.using_bitfield_type_element.basic_enable_reg,51,,1,,,1,,0,1,,,,
an_element.using_bitfield_type_element.example_prescale_reg,51,,1,,,1,,1,5,,,,
an_element.using_bitfield_type_element.res,51,,1,,Reserved bits,1,,6,2,,,,
array_of_special_type[0].element_a,52,,1,uint8_t,,1,,,,,,,
array_of_special_type[0].element_b_with_access_3,53,,4,uint32_t,,3,,,,,,,
array_of_special_type[0].res,57,11,1,uint8_t,Reserved bytes,0,11,,,,,,
array_of_special_type[1].element_a,68,,1,uint8_t,,1,,,,,,,
array_of_special_type[1].element_b_with_access_3,69,,4,uint32_t,,3,,,,,,,
array_of_special_type[1].res,73,11,1,uint8_t,Reserved bytes,0,11,,,,,,
array_of_special_type[2].element_a,84,,1,uint8_t,,1,,,,,,,
array_of_special_type[2].element_b_with_access_3,85,,4,uint32_t,,3,,,,,,,
array_of_special_type[2].res,89,11,1,uint8_t,Reserved bytes,0,11,,,,,,
longer_bitfield.basic_enable_reg,100,,2,,,1,,0,6,,,,
longer_bitfield.example_prescale_reg,100,,2,,,1,,6,7,,,,
longer_bitfield.same_offset_over_8_bit_offset,100,,2,,,1,,13,1,,,,
longer_bitfield.res,100,,2,,Reserved bits,1,,14,2,,,,
other_element,102,,2,uint16_t,,1,,,,,,,
name,offset,type_size,type,description,access
foo,0,1,uint8_t,Example of parameter not generated,1
bar,1,1,uint8_t,,1
